improver/utilities/solar.py

```diff
@@ -39,13 +38,53 @@
 from numpy import ndarray
 
 from improver import BasePlugin
+from improver.constants import HOURS_IN_DAY, MINUTES_IN_HOUR
 from improver.metadata.utilities import (
     create_new_diagnostic_cube,
     generate_mandatory_attributes,
 )
 from improver.utilities.spatial import lat_lon_determine, transform_grid_to_lat_lon
 
 
+def get_day_of_year(time: datetime) -> int:
+    """Get day of the year from given datetimes.
+
+    Args:
+        time:
+            Datetime from which to extract day-of-year.
+
+    Returns:
+        The day of year corresponding to the specified datetime.
+    """
+    if time.tzinfo is None:
+        start_of_year = datetime(time.year, 1, 1)
+    else:
+        start_of_year = datetime(time.year, 1, 1, tzinfo=timezone.utc)
+    return (time - start_of_year).days
+
+
+def get_utc_hour(time: datetime) -> float:
+    """Get utc_hour from datetime.
+
+    Args:
+        time:
+            Datetime from which to extract utc-hour.
+
+    Returns:
+        The utc hour corresponding to the specified datetime. Minute values
+        are expressed as fraction of an hour.
+    """
+    # Round times to nearest minute by adding seconds component to times
+    rounded_time = time + timedelta(seconds=time.second)
+    # Want to avoid the situation where rounding time takes us into the next day.
```

Comment:
This function is trying to encapsulate parts of: https://github.com/metoppv/improver/blob/732a6e05e94e77ea6216fde3d1c889958496d33e/improver/utilities/solar.py#L295
into separate functions that can be used in other places. I'm open to making changes here, but they will need to ensure that don't break the behaviour in the above function.

A couple of considerations on this:
* I think the logic around returning 24 rather than 0 is necessary as day-of-year is extracted separately. The answer is the same for 0 or 24 provided the correct value for day_of_year is used accordingly.
* this function is not rounding the nearest hour, rather returning the minutes portion as fractions on an hour
* handling timezones is somewhat messy as (from what gather) datetime values extracted from cubes `cube.coord("time").cell(0).point` are timezone naive. One option here is to check if values are timezone aware and convert these to UTC using `datetime.astimezone(tzinfo=timezone.utc)`, and for naive datetimes set timezone to utc using: `datetime.replace(tzinfo=timezone.utc)`. The we what we get in both cases is the utc_hour.